src/main_test.py

```diff
@@ -27,22 +28,66 @@ def _VerifyInheritance(self, routes_list, base_class):
+        continue
+
+      if not issubclass(route.handler, base_class):
+        inheritance_errors += '* %s does not inherit from %s.\n' % (
+            route.handler.__name__, base_class.__name__)
+
+    return inheritance_errors
 
   def testRoutesInheritance(self):
-    self._VerifyInheritance(main._UNAUTHENTICATED_ROUTES, handlers.BaseHandler)
-    self._VerifyInheritance(main._UNAUTHENTICATED_AJAX_ROUTES,
-                            handlers.BaseAjaxHandler)
-    self._VerifyInheritance(main._USER_ROUTES, handlers.AuthenticatedHandler)
-    self._VerifyInheritance(main._AJAX_ROUTES,
-                            handlers.AuthenticatedAjaxHandler)
-    self._VerifyInheritance(main._ADMIN_ROUTES, handlers.AdminHandler)
-    self._VerifyInheritance(main._ADMIN_AJAX_ROUTES, handlers.AdminAjaxHandler)
-    self._VerifyInheritance(main._CRON_ROUTES, handlers.BaseCronHandler)
-    self._VerifyInheritance(main._TASK_ROUTES, handlers.BaseTaskHandler)
+    errors = ''
+    errors += self._VerifyInheritance(main._UNAUTHENTICATED_ROUTES,
+                                      handlers.BaseHandler)
+    errors += self._VerifyInheritance(main._UNAUTHENTICATED_AJAX_ROUTES,
+                                      handlers.BaseAjaxHandler)
+    errors += self._VerifyInheritance(main._USER_ROUTES,
+                                      handlers.AuthenticatedHandler)
+    errors += self._VerifyInheritance(main._AJAX_ROUTES,
+                                      handlers.AuthenticatedAjaxHandler)
+    errors += self._VerifyInheritance(main._ADMIN_ROUTES,
+                                      handlers.AdminHandler)
+    errors += self._VerifyInheritance(main._ADMIN_AJAX_ROUTES,
+                                      handlers.AdminAjaxHandler)
+    errors += self._VerifyInheritance(main._CRON_ROUTES,
+                                      handlers.BaseCronHandler)
+    errors += self._VerifyInheritance(main._TASK_ROUTES,
+                                      handlers.BaseTaskHandler)
+    if errors:
+      self.fail('Some handlers do not inherit from the correct classes:\n' +
+                errors)
+
+  def testStrictHandlerMethodRouting(self):
+    """Checks that handler functions properly limit applicable HTTP methods."""
+    router = webapp2.Router(main._USER_ROUTES + main._AJAX_ROUTES +
+                            main._ADMIN_ROUTES + main._ADMIN_AJAX_ROUTES)
+    routes = router.match_routes + router.build_routes.values()
+    errors = ''
+    while routes:
+      route = routes.pop()
+      if issubclass(route.__class__, webapp2_extras.routes.MultiRoute):
+        routes += list(route.get_routes())
+        continue
+
+      if issubclass(route.handler, webapp2.RedirectHandler):
+        continue
+
+      if route.handler_method and not route.methods:
+        errors += ('* %s specifies a handler_method but no "methods" attribute, '
+                  'and may be vulnerable to XSRF via GET requests' %
```

Comment:
also add newline? or maybe something like:

failed_routes = []
(in conditional):
failed_routes.add(route.template)

and then just add ', '.join(failed_routes) ?

also -- mostly for my curiosity, does the route.template in case of a prefix route include the prefix?